packages/common/src/store/buy-crypto/sagas.ts

```diff
@@ -633,36 +659,253 @@ function* doBuyCryptoViaSol({
         userbank: userbank?.toBase58()
       }
     })
-    if (mint === 'audio') {
-      yield* call(
-        track,
-        make({
-          eventName: Name.BUY_AUDIO_FAILURE,
-          provider,
-          requestedAudio: amount,
-          error: error.message,
-          stage: BuyAudioStage.START
-        })
+    yield* call(
+      track,
+      make({
+        eventName: Name.BUY_CRYPTO_ON_RAMP_FAILURE,
+        mint,
+        provider,
+        requestedAmount: amount,
+        error: error.message
+      })
+    )
+  }
+}
+
+/**
+ * Failures happen when the on ramp purchase could not be confirmed, or the
+ * subsequent swap attempts all fail. Get the SOL balance of the wallet, and
+ * if it has a balance greater than min rent, attempt to swap all of it to the
+ * output token. If we get enough of the output token after the swap, it's a
+ * successful recovery.
+ */
+function* recoverBuyCryptoViaSolIfNecessary() {
+  yield* call(waitForAccount)
+  // Pull from context
+  const audiusBackendInstance = yield* getContext('audiusBackendInstance')
+  const getFeatureEnabled = yield* getContext('getFeatureEnabled')
+  const { track, make } = yield* getContext('analytics')
+  const reportToSentry = yield* getContext('reportToSentry')
+
+  // Return early if feature not enabled
+  if (!getFeatureEnabled(FeatureFlags.BUY_USDC_VIA_SOL)) {
+    return
+  }
+
+  // Check for local storage
+  const audiusLocalStorage = yield* getContext('localStorage')
+  const localStorageState = yield* call(
+    [
+      audiusLocalStorage,
+      audiusLocalStorage.getJSONValue<BuyCryptoViaSolLocalStorageState>
+    ],
+    BUY_CRYPTO_VIA_SOL_STATE_KEY
+  )
+
+  // Do nothing if there's no local storage state
+  if (!localStorageState) {
+    return
+  }
+
+  // Pre-emptively clear while working
+  yield* call(
+    [audiusLocalStorage, audiusLocalStorage.removeItem],
+    BUY_CRYPTO_VIA_SOL_STATE_KEY
+  )
+
+  const { mint, amount, provider, createdAt, intendedLamports } =
```

Comment:
Worth noting that if anything between here and the `try` throws an exception, we will not replace the localStorage state? My "safe enough" strategy has been to trust reading things like context values and (sort of unsafely) getting config values. But anything related to fee payer/wallets/etc has a decent number of Sentry errors which leads me to believe that we see exceptions attempting to read those values.

And yes, I feel bad for suggesting you change to this strategy and then nitpicking it afterwards 😅

Reply:
great point! I was thinking the other calls were pretty safe but better to just delete later anyway